src/Horspool.java

```diff
@@ -1,44 +1,39 @@
 import java.util.HashMap;
 import java.util.Scanner;
 
-public class Horspool {
+public class StringMatch {
 
-    private static int patternMatch(String text, String pattern) {
-        HashMap<Character,Integer> hashMap = getShiftTable(text,pattern);
+    public static int patternMatch(String text, String pattern) {
+        HashMap<Character, Integer> hashMap = getShiftTable(pattern);
         int patternLength = pattern.length();
         int textLength = text.length();
-        int i,j;
-        for(i=patternLength-1;i<textLength;i+=hashMap.get(text.charAt(i)))    {
-            for(j=0;(j<patternLength)&&(text.charAt(i-j)==pattern.charAt(patternLength-1-j));j++);
-            if(j==patternLength)
-                return i-patternLength+1;
+        int i, j;
+        Integer shift;
+        for (i = patternLength - 1; i < textLength; shift =hashMap.get(text.charAt(i)),  i += shift != null? shift: patternLength ) {
+            for (j = 0; (j < patternLength) && (text.charAt(i - j) == pattern.charAt(patternLength - 1 - j)); j++) ;
+            if (j == patternLength)
+                return i - patternLength + 1;
         }
         return -1;
     }
 
-    private static HashMap<Character,Integer> getShiftTable(String text,String pattern)  {
-        HashMap<Character,Integer> hashMap  = new HashMap<>();
+    private static HashMap<Character, Integer> getShiftTable(String pattern) {
+        HashMap<Character, Integer> hashMap = new HashMap<>();
         int len = pattern.length();
-        int textLength = text.length();
-
-        for(int i=0;i<textLength;i++)
-            if(!hashMap.containsKey(text.charAt(i)))
-                hashMap.put(text.charAt(i),len);
-
-        for(int i=0;i<len-1;i++)
-            hashMap.replace(pattern.charAt(i),len-1-i);
+        for (int i = 0; i < len - 1; i++)
+            hashMap.put(pattern.charAt(i), len - 1 - i);
```

Comment:
`HashMap.put` method puts the key-value pair in the Map if it was not exsisting. If the key exsisted, it simply updates the value of key.